query-parser/src/parse.rs

```diff
@@ -83,320 +96,348 @@ error_chain! {
+
+                    // Pattern::new takes care of reversal of reversed
+                    // attributes: [?x :foo/_bar ?y] turns into
+                    // [?y :foo/bar ?x].
+                    //
+                    // This is a bit messy: the inner conversion to a Pattern can
+                    // fail if the input is something like
+                    //
+                    // ```edn
+                    // [?x :foo/_reversed 23.4]
+                    // ```
+                    //
+                    // because
+                    //
+                    // ```edn
+                    // [23.4 :foo/reversed ?x]
+                    // ```
+                    //
+                    // is nonsense. That leaves us with a nested optional, which we unwrap here.
+                    Pattern::new(src, e, a, v, tx)
+                        .map(WhereClause::Pattern)
+                        .ok_or(combine::primitives::Error::Expected("pattern".into()))
+                }))
 });
 
-def_value_parser_fn!(Where, clause, WhereClause, input, {
-    choice([Where::pattern(),
-            Where::pred(),
+def_parser!(Where, clause, WhereClause, {
+    choice([try(Where::pattern()),
             // It's either
             //   (or-join [vars] clauses…)
             // or
             //   (or clauses…)
             // We don't yet handle source vars.
-            Where::or_join_clause(),
-            Where::or_clause(),
-    ]).parse_stream(input)
+            try(Where::or_join_clause()),
+            try(Where::or_clause()),
+
+            try(Where::pred()),
+    ])
 });
 
-def_value_parser_fn!(Where, clauses, Vec<WhereClause>, input, {
+def_parser!(Where, clauses, Vec<WhereClause>, {
     // Right now we only support patterns and predicates. See #239 for more.
     (many1::<Vec<WhereClause>, _>(Where::clause()))
-    .skip(eof())
-    .parse_stream(input)
 });
 
-pub struct Find<I>(::std::marker::PhantomData<fn(I) -> I>);
+pub struct Find;
 
-def_value_parser_fn!(Find, period, (), input, {
-    matches_plain_symbol!(".", input)
+/// TODO: extract macro for matching these `PlainSymbol` instances.
+def_parser!(Find, period, edn::ValueAndSpan, {
+    satisfy(|v: edn::ValueAndSpan| {
+        if let edn::SpannedValue::PlainSymbol(ref s) = v.inner {
+            s.0.as_str() == "."
+        } else {
+            false
+        }
+    })
 });
 
-def_value_parser_fn!(Find, ellipsis, (), input, {
-    matches_plain_symbol!("...", input)
+def_parser!(Find, ellipsis, edn::ValueAndSpan, {
+    satisfy(|v: edn::ValueAndSpan| {
+        if let edn::SpannedValue::PlainSymbol(ref s) = v.inner {
+            s.0.as_str() == "..."
+        } else {
+            false
+        }
+    })
 });
 
-def_value_parser_fn!(Find, find_scalar, FindSpec, input, {
+def_parser!(Find, find_scalar, FindSpec, {
     Query::variable()
+        .map(|var| FindSpec::FindScalar(Element::Variable(var)))
         .skip(Find::period())
```

Comment:
I don't think this is in the right place. You should skip the `period` before the `.map` — you're matching 

```edn
[:find ?x . :in …]
```

In this order, we'll construct the `FindScalar` instance, then fail to match the period in the failure case, which isn't free.